### PR TITLE
refactor macro with AMP

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes #ISSUE_NUMBER
as the tiltle, optimize the macro with AMP and put the macro in `.hpp` file, so that we can use it for custom device.  @bdhirsh  @albanD  
as we talked at this discuss, optimize the macros so that we can add a new macro for other devide, and move these macros to `.hpp` so that we can include these macros with custom device to configure the ops. 
https://dev-discuss.pytorch.org/t/improve-the-extension-with-privateuse1-for-custom-device/1196/7 
cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Comment by heidongxianhua** [comment]
> optimize the macros so that we can add a new macro for other devide, and move these macros to `.hpp` so that we can include these macros with custom device to configure the ops.

yeah, sorry.  optimize the macros so that we can add a new macro for other devide, and move these macros to .hpp so that we can include these macros with custom device to configure the ops.

> Could you explain in the PR description exactly what you've done please? I guess some rename + moving things around?

yeah, sorry without clearly described. and your guess is correct. I want to optimize the macros so that we are easier to add a new macro for a new dvice, and move these macros to `.hpp` so that we can include it  to configure ops for custom device. @albanD

**Comment by bdhirsh** [comment]
From a quick look around: other than moving some macros from the cpp file into headers, the major changes I see are macro renaming:
```
KERNEL -> KERNEL_CPU, KERNEL_CUDA, KERNEL_PRIVATEUSEONE
KERNEL2 -> KERNEL_CPU2, KERNEL_CUDA2, KERNEL_PRIVATEUSEONE2
KERNEL_DIFFERENT_REDISPATCH_SIGNATURE -> ... CPU, CUDA, PRIVATEUSE1
```

is that right?

**Comment by heidongxianhua** [comment]
> From a quick look around: other than moving some macros from the cpp file into headers, the major changes I see are macro renaming:
> 
> ```
> KERNEL -> KERNEL_CPU, KERNEL_CUDA, KERNEL_PRIVATEUSEONE
> KERNEL2 -> KERNEL_CPU2, KERNEL_CUDA2, KERNEL_PRIVATEUSEONE2
> KERNEL_DIFFERENT_REDISPATCH_SIGNATURE -> ... CPU, CUDA, PRIVATEUSE1
> ```
> 
> is that right?
yes,  that is it.

**Review by albanD — APPROVED** [review]
Sounds good to me!

File: torch/amp/autocast_mode.py [hunk 1]
```diff
@@ -320,8 +320,8 @@ def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any):  # type: ignore[ov
         elif self.device == self.custom_backend_name:
             if torch.autocast_decrement_nesting() == 0:
                 torch.clear_autocast_cache()
-            self.custom_device_mod.set_autocast_enabled(self._enabled)
-            self.custom_device_mod.set_autocast_dtype(self.fast_dtype)
+            self.custom_device_mod.set_autocast_enabled(self.prev)
```
[the remark this answers is not in the document]
Comment (heidongxianhua) [reply torch/amp/autocast_mode.py:323]: > This is a bugfix from the previous PR unrelated to this change right?

yes, it is a bugfix.

**Review by bdhirsh — APPROVED** [review]

**Pull request closed without merging**